File: app/src/main/java/com/b18/kipsafe/SharedPreferenceManager.java
package com.b18.kipsafe;

import android.content.Context;
import android.content.SharedPreferences;

import java.text.ParseException;
import java.util.Calendar;

import static com.b18.kipsafe.converters.IsoConverterKt.convertIsoToCalendar;

/**
 * Manage SharedPreferences.
 */
@Deprecated
public class SharedPreferenceManager {

    private SharedPreferences prefs;
    private Context context;

    public SharedPreferenceManager(Context context) {
        prefs = android.preference
                .PreferenceManager.getDefaultSharedPreferences(context);
        this.context = context;
    }

    /**
     * Save the number of minutes before sunset when the alarm should go off.
     *
     * @param time integer
     */
    public void writePrefTime(int time) {
        SharedPreferences.Editor edit = prefs.edit();
        String keyString = context.getResources().getString(R.string.pref_time);
        edit.putInt(keyString, time);
        edit.apply();
    }

    /**
     * Get the number of minutes before sunset when the alarm should go off.
     *
     * @return time integer
     */
    public int getPrefTime() {
        String keyString = context.getResources().getString(R.string.pref_time);
        return prefs.getInt(keyString, -1);
    }

    /**
     * Save the last known sunset, in case there is no internet when an alarm wants to reset.
     * @param time ISO string
     */
    public void saveSunsetTime(String time) {
        SharedPreferences.Editor edit = prefs.edit();
        String keyString = context.getResources().getString(R.string.pref_sunset);
        edit.putString(keyString, time);
        edit.apply();
    }

    /**
     * Get the last known sunset, default time if could not parse api result.
     * @return Calendar.
     * @throws DataNotFoundException when the last known sunset could not be found
     */
    public Calendar getSunsetTime() throws DataNotFoundException {
        String keyString = context.getResources().getString(R.string.pref_sunset);
        String defaultTime = "2017-10-11T17:00:00+00:00";
        String time = prefs.getString(keyString, defaultTime);
        if (time.equals(defaultTime)) {
            throw new DataNotFoundException("Could not find requested data in SharedPreferences");
        }
        try {
            return convertIsoToCalendar(time);
        } catch (ParseException e) {
            e.printStackTrace();
        }
        // Couldn't understand what the sunset api returned.
        Calendar calendar = Calendar.getInstance();
        calendar.set(Calendar.HOUR_OF_DAY, 18);
        return calendar;
    }

    /**
     * Combines sunset and preferred minutes before sunset to an alarm time.
     * @return preferred alarm time
     * @throws DataNotFoundException when the last known sunset could not be found
     */
    public Calendar getAlarmTime() throws DataNotFoundException {
        Calendar time = getSunsetTime();
        int minutes = getPrefTime();
        time.add(Calendar.MINUTE, -minutes);
        return time;
    }

    /**
     * Save boolean whether the alarm is set or not.
     */
    public void saveIsAlarmSet(boolean isAlarmSet) {
        SharedPreferences.Editor edit = prefs.edit();
        String keyString = context.getResources().getString(R.string.pref_isalarmset);
        edit.putBoolean(keyString, isAlarmSet);
        edit.apply();
    }

    /**
     * Get whether the alarm is set or not.
     * @return true if alarm is set
     */
    public boolean getIsAlarmSet() {
        String keyString = context.getResources().getString(R.string.pref_isalarmset);
        return prefs.getBoolean(keyString, false);
    }

}

File: app/src/main/java/com/b18/kipsafe/SharedPreferenceManager.kt
package com.b18.kipsafe

import android.content.Context
import android.content.SharedPreferences
import com.b18.kipsafe.converters.convertIsoToCalendar
import com.b18.kipsafe.util.prefIsAlarmSet
import com.b18.kipsafe.util.prefSunset
import com.b18.kipsafe.util.prefTime
import java.text.ParseException
import java.util.*

class SharedPreferenceManager(val context: Context) {
    private val preferences: SharedPreferences = context.getSharedPreferences("kip", Context.MODE_PRIVATE)

    // Actually property getters and setters should not be used when accessing slow things like shared preferences, but not a big performance issue here.

    /** The number of minutes before sunset to set the alarm. */
    var minutesBeforeSunset: Int
        get() = preferences.getInt(UserPreference.TIME.keyString, -1)
        set(value) {
            preferences.edit().putInt(UserPreference.TIME.keyString, value).apply()
        }

    /** True if the alarm is set, false otherwise. */
    var isAlarmSet: Boolean
        get() = preferences.getBoolean(UserPreference.IS_ALARM_SET.keyString, false)
        set(value) {
            preferences.edit().putBoolean(UserPreference.IS_ALARM_SET.keyString, value).apply()
        }

    /** Last known sunset. */
    var sunset: String
        get() = preferences.getString(UserPreference.SUNSET.keyString, "")!!
        set(value) {
            preferences.edit().putString(UserPreference.SUNSET.keyString, value).apply()
        }

    /** Whether the alarm should go of during weekends only. */
    var isWeekendOnly: Boolean
        get() = preferences.getBoolean(UserPreference.WEEKEND_ONLY.keyString, false)
        set(value) {
            preferences.edit().putBoolean(UserPreference.WEEKEND_ONLY.keyString, value).apply()
        }

    /**
     * Get the last known sunset. This will be 18:00 hrs if we cannot parse api results.
     *
     * @throws DataNotFoundException when the last known sunset could not be found.
     */
    fun getSunset(): Calendar {
        // Copy value so we only access shared prefs once.
        val time = sunset
        if(time == "") throw DataNotFoundException(
                "Could not find requested time in SharedPreferences.")

        return try {
            convertIsoToCalendar(time)
        } catch (e: ParseException) {
            val calendar = Calendar.getInstance()
            calendar.set(Calendar.HOUR_OF_DAY, 18)
            e.printStackTrace()
            calendar
        }
    }

    /**
     * Combines sunset and preferred minutes before sunset to return an alarm time.
     *
     * @return Calendar object with preferred allar
     */
    fun getAlarmTime(): Calendar {
        val time = getSunset()
        time.add(Calendar.MINUTE, -minutesBeforeSunset)
        return time
    }

}